common/src/main/java/org/corfudb/common/protocol/client/ChannelHandler.java

```diff
@@ -0,0 +1,315 @@
+            errorOutAllPendingRequests();
+            CompletableFuture cf = channelCf;
+            if (!cf.isDone() && !cf.isCompletedExceptionally()) {
+                // throw shutdown exception instead ?
+                cf.completeExceptionally(new PeerUnavailable(remoteAddress));
+            }
+
+            Channel currentChannel = channel;
+            if (currentChannel != null) {
+                // TODO(Maithem): will this end up calling channelInactive? need to cancel timer task
+                currentChannel.close().addListener(res -> {
+                    if (!res.isSuccess()) {
+                        log.warn("Failed to close channel for {}", res.cause());
+                    }
+                });
+            }
+        } finally {
+            requestLock.writeLock().unlock();
+        }
+    }
+
+
+    private ChannelInitializer getChannelInitializer() {
+        return new ChannelInitializer() {
+            @Override
+            protected void initChannel(@Nonnull Channel ch) throws Exception {
+                ch.pipeline().addLast(new IdleStateHandler(config.getIdleConnectionTimeoutInMs(),
+                        config.getKeepAlivePeriodInMs(), 0));
+
+                ch.pipeline().addLast(new LengthFieldPrepender(4));
+                ch.pipeline().addLast(new LengthFieldBasedFrameDecoder(Integer.MAX_VALUE,
+                        0, 4, 0,
+                        4));
+
+                /**
+                 ch.pipeline().addLast(new NettyCorfuMessageDecoder());
+                 ch.pipeline().addLast(new NettyCorfuMessageEncoder());
+                 **/
+
+                /**
+                 * //TODO(Maithem): need to implement new handshake logic without netty pipelines
+                 ch.pipeline().addLast(new ClientHandshakeHandler(parameters.getClientId(),
+                 node.getNodeId(), parameters.getHandshakeTimeout()));
+                 **/
+
+                ch.pipeline().addLast(ChannelHandler.this);
+            }
+        };
+    }
+
+    protected long generateRequestId() {
+        return idGenerator.incrementAndGet();
+    }
+
+    //TODO(Maithem) Add keep-alive logic here
+
+    private void checkRequestTimeout() {
+        while (!requestTimeoutQueue.isEmpty()) {
+            RequestTime request = requestTimeoutQueue.peek();
+            if (request == null || (System.currentTimeMillis() - request.creationTimeMs) < config.getRequestTimeoutInMs()) {
+                // if there is no request that is timed out then exit the loop
+                break;
+            }
+            request = requestTimeoutQueue.poll();
+            CompletableFuture requestFuture = pendingRequests.remove(request.requestId);
+            if (requestFuture != null && !requestFuture.isDone()) {
+                requestFuture.completeExceptionally(new PeerUnavailable(remoteAddress));
+            } else {
+                // request is already completed successfully.
+            }
+        }
+    }
```

Comment:
I haven't started reviewing, just took a glance at the class, so I might be wrong.
It seems that using a Java DelayQueue for checking timeout request is simplier and more effiecnt here. The efficiency comes from the fact that current scheduling impl might have unnecessary checks when a request is just issued, and the checked is immediately scheduled, while the delayqueue blocks until the first request expires. 

Also I'm not understanding why we use eventLoopGroup to drive the scheduled checking here? Is it better to use a separate thread?

Reply:
Looked at it again and I think the problem with current scheduling approach is not extra check, it's not fully repecting the request timeout, the real timeout could be 2x the configured. Since the check interval is same as reqeust timeout, we could face the problem that after the request is put into the queue, right before it times out, we scheduled a check, but think that it's not timed out yet, and wait to to be scheduled next time. However by then the time elapsed is almost 2x the configured request timeout.

Reply:
The reason I didn't use a new thread is simply because the task to "too cheap".

Good catch. I think one way to fix this is to lower the timer resolution for the check. Like check every 100ms/200ms for example. I wanted an easy way to cancel the pending requests (before the timeout expires) that's why I didn't use the scheduled ScheduledExecutorService. Its possible to use the ScheduledExecutorService in conjunction with tracking the returned future task and cancel them from the scheduler, but that uses more than double the memory?

Reply:
@Maithem The way I was thinking is to use a delayQueue, and a dedicated thread to block waiting on the head of the queue which is the first one to be timed out. This involves a spearate thread, which may have some overhead, as this is per channel. I think the approach that the current NettyClientRouter uses might be better, it uses `CFUtils.with()`, which internally uses a global (static) single thread ScheduledExecutorService which is invoked only when the request is timed out (similar to DelayQueue, and actually ScheduledExecutorService internally uses DelayQueue). This approach only requires one extra thread that can monitor all the channels, which is much cheaper.

Reply:
But how can you cancel a task without the timer expiring ?

You need to capture the task future returned from ScheduledExecutorService and store it somewhere, which doubles the memory. Since the tasks are being stored inside the executor and also inside your other data structure.

Reply:
You mean cancel a task before it times out like when channel inactive/disconnected? If this is the case, then  `CFUtils.within()` can still do it, when `sendRequest()` is called, you first generate a task future as usual, put this task future into pendingRequest map, and return to client a composed future, i.e. `return CFUtils.within(task, timeout)`. When you need to cancel a task, just cancel the orignal task in the pendingRequest map, and this composed future will be canceled as well.

Reply:
Yep. There are situations, where the future can be canceled before the time elapses. This is required to let upper layers know quickly, so that they can retry, or make other decisions (i.e. Fault detector). 

That's what I was saying, you need another queue/collection to keep the pending a task future in addition to the structures that are maintained inside the ScheduledExecutorService. My point is why have two mechanisms to expire a request, when you can only have one ?

Also, this reduces object allocation and takes less resources. The current code only needs one append to the queue + poller thread. The ScheduledExecutorService maintains an internal min heap that it needs to maintain. Furthermore, I think the ScheduledExecutorService doesn't release the task if you cancel it until the timer kicks in, so It wont be GC'd as fast. I think that's the default behavior.

Reply:
@Maithem Sorry I didn't get the exrtra collection part, what I was referring to actually reduces explicit collecitons. The CFUtil.within() approach only requires a explicit pendingRequest map, which is already there (plus an implicit queue in the scheduler), while in your current code, you have a pendingRequest map as well as requestTimeoutQueue. So in terms of collections, it's actually the same, while the problem in current implementation is it wastes cpu cycles to check timout every 100~200 ms in order to repect the configured timeout, is that right?